Import React Concurrent Mode Profiler

taneliang opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test-prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

This PR imports the concurrent mode profiler [project](https://github.com/MLH-Fellowship/scheduling-profiler-prototype) into `packages/react-devtools-scheduling-profiler`, with history 🎉 

![image](https://user-images.githubusercontent.com/12784593/90470142-47b50b00-e14d-11ea-86dc-479cc51cdb2f.png)

Broadly follows the approach of https://github.com/facebook/react/pull/16381.

Resolves https://github.com/MLH-Fellowship/scheduling-profiler-prototype/issues/134.

cc @bvaughn  

## Import Details

### Notes

* Our new `build` and `start` scripts were failing with this error: https://github.com/babel/babel/issues/11216#issue-576326728. I've followed the maintainer's recommendation and **updated all `@babel/*` packages across the repository**, excluding fixtures. We can easily revert this (i.e. https://github.com/facebook/react/pull/19634/commits/babe03cedd678e72d1caf54d889d930ad8b293a2) if there's a better solution.
* Webpack config is based on react-devtools-shell's.
* Dependencies used have been upgraded to the latest versions. `css-loader` config differences can be backported into the other devtools packages later.
* Unlike the other devtools packages, we generate our `index.html` file into the `dist` folder using `HtmlWebpackPlugin`. This means that the `dist` folder can be served directly, and can likely be deployed to Vercel without a need for a `now.json` file.

Major differences with the version in the original repo ([full diff](https://github.com/MLH-Fellowship/react/compare/4aa36c1...MLH-Fellowship:scheduling-profiler-import)):

* Parcel has been replaced with Webpack for better integration with the rest of the React repo.
* The sample profile auto-loaded in DEV for convenience is now dynamically imported. It's now trivial to add a demo profile feature for end users (https://github.com/MLH-Fellowship/scheduling-profiler-prototype/issues/82).
* index.html is now automatically generated by html-webpack-plugin instead of being based on a template file.
* Page `title` has been renamed to "React Concurrent Mode Profiler".

### Process

Similar to facebook/react#16381, the goal of this import is to preserve history from https://github.com/MLH-Fellowship/scheduling-profiler-prototype. Verified that `git log --follow` worked as expected on imported files.

#### Prepare fork for merge

```sh
mkdir profiler-merge
cd profiler-merge

git init .
git commit -m "Initializing empty merge repo" --allow-empty  

git remote add -f profiler-merge git@github.com:MLH-Fellowship/scheduling-profiler-prototype.git
git merge profiler-merge/master --allow-unrelated-histories

mkdir -p packages/react-scheduling-profiler
mv * packages/react-devtools-scheduling-profiler
mv .* packages/react-devtools-scheduling-profiler
mv packages/react-devtools-scheduling-profiler/.git .
# Commit
```

#### Import into React

```sh
cd ../react

git remote add profiler-importable ../profiler-merge
git fetch --all

git checkout -b scheduling-profiler-import
git merge --allow-unrelated-histories profiler-importable/importable
git remote rm profiler-importable
```

## Checklist

- [x] Remove extraneous project-specific files (.github, .vscode, etc)
- [x] Integrate package.json with monorepo and remove yarn.lock
- [x] Fix build/start
- [x] Integrate with linter
- [x] Fix lint
- [x] Integrate with Flow
- [x] Fix Flow types, if required
- [x] Integrate with tests
- [x] Fix tests
- [x] Add FB copyright header
- [x] Fix CI

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->
* CI
* In react-scheduling-profiler package:
	* Test prod build
		1. `yarn build`
			![image](https://user-images.githubusercontent.com/12784593/90470702-b6df2f00-e14e-11ea-87ab-37ef90fb98d8.png)
		1. `cd dist`
		1. `serve` (using https://www.npmjs.com/package/serve)
			![image](https://user-images.githubusercontent.com/12784593/90471015-96fc3b00-e14f-11ea-97e4-24a7f9959377.png)
	* Test development:
		1. `yarn start` works
			![image](https://user-images.githubusercontent.com/12784593/90470964-6ddbaa80-e14f-11ea-9d7f-d49ad96b24ea.png)
		1. Test hot reloading works

bvaughn commented:
> * Our new `build` and `start` scripts were failing with this error: https://github.com/babel/babel/issues/11216#issue-576326728. I've followed the maintainer's recommendation and **updated all `@babel/*` packages across the repository**, excluding fixtures. We can easily revert this (i.e. https://github.com/facebook/react/pull/19634/commits/babe03cedd678e72d1caf54d889d930ad8b293a2) if there's a better solution.

I know this is kind of a pain, but can you manually diff the artifacts of the other packages (e.g. react, react-dom, scheduler, react-devtools*, etc) to ensure that there are no changes as a result of this update? Need to make sure we don't have any regressions.

> * Dependencies used have been upgraded to the latest versions. `css-loader` config differences can be backported into the other devtools packages later.

Can you clarify? Does this mean other DevTools packages are going to be broken in the meanwhile?

bvaughn commented:
Let's maybe not commit the Chrome sample profile. It's pretty huge (18.2 MB) and I'd rather not bloat the Git checkout. Can we strip it entirely so it's not in the history either?

taneliang commented:
> can you manually diff the artifacts of the other packages (e.g. react, react-dom, scheduler, react-devtools*, etc) to ensure that there are no changes as a result of this update? 

Sure, I'll do a build on both master and this branch and see what's changed. Is there anything I should look out for in particular, or should I just check that nothing changed at all?

> Can you clarify? Does this mean other DevTools packages are going to be broken in the meanwhile?

Oh no, I meant that we're using the latest version of `css-loader`, which is a few major versions ahead of the `css-loader` used in the other DevTools packages. One of these major versions changed the Webpack config options, so the `css-loader` config that we're using is slightly different (just a 2 line change) from the one in the other DevTools packages.

We can replicate the config updates over to the other DevTools packages' Webpack configs if/when we want to upgrade `css-loader` there.

I just wanted to highlight that the package versions that the scheduling profiler is using is different from the ones used by the other DevTools packages.

> Also, what's left to do to get CI auto-deploying this app?

Ah I haven't looked into that. It really should just be running `yarn build` and uploading the resulting `dist` folder to Vercel/Netlify. Is something similar done by CI for the other DevTools sites? Also, should we do this in this PR or should we work on that once this is landed?

> Anything I can help with?

Yeah, would you want to create the site on Vercel? I'll be more than happy to deploy it onto my personal Vercel account, but I'm not sure if you'll be comfortable with that 😆 

> Let's maybe not commit the Chrome sample profile. It's pretty huge (18.2 MB) and I'd rather not bloat the Git checkout. Can we strip it entirely so it's not in the history either?

Sure! I'll remove it tomorrow, and I'll try to remove all the other sample files in the history too.

bvaughn commented:
> Sure, I'll do a build on both master and this branch and see what's changed. Is there anything I should look out for in particular, or should I just check that nothing changed at all?

Ideally, no change at all :smile: If there are changes, then let's review them carefully and make sure they wouldn't cause any regressions.

> Oh no, I meant that we're using the latest version of css-loader, which is a few major versions ahead of the css-loader used in the other DevTools packages.

Oh, ok gotcha. I misunderstood.

> Ah I haven't looked into that. It really should just be running yarn build and uploading the resulting dist folder to Vercel/Netlify. Is something similar done by CI for the other DevTools sites? Also, should we do this in this PR or should we work on that once this is landed?

Gotcha. We can follow up on this then. This PR is huge enough. :grin:

> Sure! I'll remove it tomorrow, and I'll try to remove all the other sample files in the history too.

Great! Thanks. I'll take a look through the rest of the code in the meanwhile.

kartikcho commented:
@bvaughn should I start manually migrating open issues too or wait until after this has landed?

taneliang commented:
> Sure! I'll remove it tomorrow, and I'll try to remove all the other sample files in the history too.

Done! What I did:

1. Ran [this `git-show-big-files` script](https://gist.github.com/mcxiaoke/b4bdad5727c9400bbf7d101f27297e86) to check what the largest files were:

		[I] ~/react (scheduling-profiler-import|…2) ./git-show-big-files.sh
		All sizes are in kB's. The pack column is the size of the object, compressed, inside the pack file.
		size   SHA                                       location
		18625  593fcad24169745bffe7ddefc9be87f0183ea0f7  packages/react-scheduling-profiler/static/sample-chrome-profile.json
		5525   a442fbbd9a657a04777e560638023ef0b566f953  docs/img/blog/steve_reverse.gif
		4214   eeac18b9692089ca89b763cdc711ba8d54b1509f  static/perfprofilev0.json
		2691   e216017a07cf1c18e5c0354c368672fffc2e476e  static/perfprofilev1.json
		...

1. Ran BFG:
	```sh
	bfg --delete-files '{sample-chrome-profile,Profile-*,perfprofilev*,profile-filtered*,small-devtools}.json' --no-blob-protection react
	```
	
	Output:

		Deleted files
		-------------

        Filename                       Git id
        --------------------------------------------------
        Profile-20200625T133129.json | e216017a (2.6 MB)
        Profile-20200625T222157.json | 41b7784c (767.1 KB)
        Profile-20200625T224727.json | b013eff8 (1.1 MB)
        perfprofilev0.json           | eeac18b9 (4.1 MB)
        perfprofilev1.json           | e216017a (2.6 MB)
        perfprofilev2.json           | 593fcad2 (18.2 MB)
        profile-filtered.json        | c0a8766f (44.0 KB)
        profile-filtered2.json       | 6901f5cf (1.4 KB)
        profile-filtered3.json       | 8dcffc71 (7.0 KB)
        sample-chrome-profile.json   | 593fcad2 (18.2 MB)
        small-devtools.json          | eeac18b9 (4.1 MB)

1. `git reflog expire --expire=now --all && git gc --prune=now --aggressive`

1. Ran `git-show-big-files` to confirm that the big files have been deleted:

	```
	[I] ~/react (scheduling-profiler-import↑91↓91|…2) ./git-show-big-files.sh
	All sizes are in kB's. The pack column is the size of the object, compressed, inside the pack file.
	size  SHA                                       location
	5525  a442fbbd9a657a04777e560638023ef0b566f953  docs/img/blog/steve_reverse.gif
	1835  245868f6f454cb696323a9e8aa5cfb68fa7418ab  fixtures/dom/public/test.mp4
	1783  7b4405f62e7caab5c42d1fdd92cef0b4443cac7e  docs/img/blog/cra-runtime-error.gif
	1695  88db2896fa46cdc0bb2cdc3c5f307ceb73613103  docs/img/blog/cra-dynamic-import.gif
	1378  fd7ed94938f188ebb73138331b0fd27772173f96  docs/img/blog/devtools-full.gif
	...
	```

A fresh clone of the MLH-Fellowship/react repository will still contain the removed files as GitHub hasn't gced them. I hope this won't be a problem for us since the commit history that will be merged won't contain them?

I've also removed the code that automatically loads a sample profile in DEV. It's a really nice convenience feature that we probably will want to re-add after this PR is landed.

taneliang commented:
> Ideally, no change at all 😄 If there are changes, then let's review them carefully and make sure they wouldn't cause any regressions.

There are about 2300 lines of changes in the main React packages, but I didn't notice anything concerning. I've uploaded the diffs here: https://gist.github.com/taneliang/c193548a1d5f07f083486ce859c2edf1

Interestingly, it seems like I can't build any of the devtools packages on master without encountering the same error that caused me to upgrade Babel in the first place: `Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main resolved in /home/e-liang/react/node_modules/@babel/helper-compilation-targets/package.json`. I've tried the latest versions of both Node 12 and Node 14. It still seems to work in Node 10, so @kartikcho will generate the diff for that in a bit.

What I did:

Main packages:

```sh
git checkout scheduling-profiler-import
yarn

RELEASE_CHANNEL=experimental yarn build
cp -r build ~/react-build-spi-experimental

RELEASE_CHANNEL=stable yarn build
cp -r build ~/react-build-spi-stable

git checkout master
yarn

RELEASE_CHANNEL=experimental yarn build
cp -r build ~/react-build-master-experimental

RELEASE_CHANNEL=stable yarn build
cp -r build ~/react-build-master-stable

cd ~
diff -ENwbur react-build-master-experimental react-build-spi-experimental > experimental.diff
diff -ENwbur react-build-master-stable react-build-spi-stable > stable.diff
# Both diff files uploaded to https://gist.github.com/taneliang/c193548a1d5f07f083486ce859c2edf1
```

react-devtools-* packages:

```sh
cd react
git checkout scheduling-profiler-import
yarn
yarn build-for-devtools

cd packages/react-devtools-core && rm -rf dist && yarn build
cd ../react-devtools-extensions && rm -rf build && yarn build
cd ../react-devtools-inline && rm -rf dist && yarn build
cd ../react-devtools-shell && rm -rf dist && yarn build
cd ../..
mkdir ~/react-build-spi-devtools
cp -r packages/react-devtools-core/dist ~/react-build-spi-devtools/core
cp -r packages/react-devtools-extensions/build ~/react-build-spi-devtools/extensions
cp -r packages/react-devtools-inline/dist ~/react-build-spi-devtools/inline
cp -r packages/react-devtools-shell/dist ~/react-build-spi-devtools/shell

git checkout master
yarn
yarn build-for-devtools

cd packages/react-devtools-core && rm -rf dist && yarn build
cd ../react-devtools-extensions && rm -rf build && yarn build
cd ../react-devtools-inline && rm -rf dist && yarn build
cd ../react-devtools-shell && rm -rf dist && yarn build
cd ../..
mkdir ~/react-build-master-devtools
cp -r packages/react-devtools-core/dist ~/react-build-master-devtools/core
cp -r packages/react-devtools-extensions/build ~/react-build-master-devtools/extensions
cp -r packages/react-devtools-inline/dist ~/react-build-master-devtools/inline
cp -r packages/react-devtools-shell/dist ~/react-build-master-devtools/shell

cd ~
diff -ENwbur react-build-master-devtools react-build-spi-devtools > devtools.diff
# @kartikcho is generating this
```

bvaughn commented:
@taneliang Which OS are you on?

I'm able to build DevTools packages in master without any problems using v12.16.2.

kartikcho commented:
Here's the [generated diff](https://gist.github.com/kartikcho/ffaeb0e1df859e898d5465c31f98f680) for react-devtools as discussed in the comment above. The diff seems to be minified so if that doesn't provide enough information, I'll run the commands again.

taneliang commented:
> Which OS are you on? I'm able to build DevTools packages in master without any problems using OS X and v12.16.2.

I'm using Ubuntu 18.04 in WSL2 on Windows 10, running Node v12.18.3. According to https://github.com/babel/babel/issues/11216#issuecomment-595393641 on the Babel issue thread, it seems like only later versions of Node will see this error

bvaughn commented:
I agree those changes to the React packages all look safe (either just whitespace changes or small improvements).

bvaughn commented:
> > Which OS are you on? I'm able to build DevTools packages in master without any problems using OS X and v12.16.2.
> 
> I'm using Ubuntu 18.04 in WSL2 on Windows 10, running Node v12.18.3. According to https://github.com/babel/babel/issues/11216#issuecomment-595393641 on the Babel issue thread, it seems like only later versions of Node will see this error

Gotcha. I see the error too if I build with Node v14.

bvaughn commented:
> Because this already seems to be broken on master, would you want to do this Babel upgrade in a different PR?

Yes. If the only reason to do this upgrade was to fix DevTools for later Node versions, and that didn't actually work– let's move the deps upgrade to its own PR?

Sorry to ask you to do this. It would safer though, seeing as how we're close to cutting a new release. Less chance of this merge getting reverted.

taneliang commented on packages/react-devtools-scheduling-profiler/src/ImportPage.js:
```diff
@@ -0,0 +1,128 @@
+  );
+
+  useEffect(() => {
+    if (__DEV__) {
```
> I wonder if we might not be able to do something more clever for DEV like just add a special state to the App component that manages the auto-load (so we don't show the ImportPage UI at all?)

Ooh, great suggestion!

> This could even be URL param driven (in case we want to actually load different profiles, or not load one at all so we could iterate on the ImportPage UI.)

This sounds like something that can be used to support sharing profiles too (https://github.com/MLH-Fellowship/scheduling-profiler-prototype/issues/75)! What I have in mind is a way to allow the user to paste a URL to a raw profile (e.g. the link to a raw GitHub Gist or Paste) and have the profiler load the profile directly from that link. If this link can be specified through a URL param, it can be used both for convenience in DEV and for sharing the profile + profiler with others. I'm not sure if this will be possible due to CORS rules though

bvaughn replied:
Yeah, sharing profiles sounds really nice.

There are two ways (at least) that we could do it:
* Pass the URL of a profile JSON that the app could load (may have CORS issues though)
* Pass a URL-encoded string of the actual profile data (maybe too big– but maybe we can compress things, remove bits of the profile we don't care about, etc – or maybe we can export a different format that's more streamlined with only the stuff we need)

taneliang commented:
> let's move the deps upgrade to its own PR? Sorry to ask you to do this.

No worries! I had planned for that so the Babel upgrade commit could be cherry picked and reverted easily 😆 The PR is opened at https://github.com/facebook/react/pull/19647. I'll revert commit `a633ac3` in this PR once it's merged.

taneliang commented on packages/react-devtools-scheduling-profiler/src/EventTooltip.css:
```diff
@@ -0,0 +1,79 @@
+    transparent
+  );
+  white-space: pre;
+  --gradient-height: 5em;
```
Oh, you used that to add a fade to the bottom half of taller component stacks:

|With this variable|If this line is deleted|
|-|-|
|![image](https://user-images.githubusercontent.com/12784593/90651206-5562b180-e26f-11ea-8ced-debd8857616a.png)|![image](https://user-images.githubusercontent.com/12784593/90651251-63183700-e26f-11ea-854e-4dc487f7b944.png)|

I think it's a variable because it's used in 2 places (i.e. both a prefixed and not prefixed version of `mask-image`). I'll be happy to inline it or remove the gradient entirely

bvaughn replied:
Yeah, I think since it's only used in two places (also in this file) we should just inline it. That's what my comment was trying to say, rather than asking what the CSS style itself was being used for :grin:

taneliang replied:
Oops, my bad! 😅 I'll inline this

taneliang commented on packages/react-devtools-scheduling-profiler/src/ImportPage.css:
```diff
@@ -0,0 +1,206 @@
+  right: 0;
+}
+
+/* Button2 styling */
```
@kartikcho would you want to take this up?

taneliang commented on packages/react-devtools-scheduling-profiler/src/utils/getBatchRange.js:
```diff
@@ -0,0 +1,47 @@
+import type {BatchUID, Milliseconds, ReactProfilerData} from '../types';
+
+// TODO: Batch duration probably shouldn't include delayed passive effects?
+// It should probably end with the layout effect.
```
Actually this comment was from your prototype and we didn't touch it 😆 I've removed it.

Could I clarify what you mean by delayed passive effects? Are these passive effects that are executed some time after the commit is complete?

I'm asking this because this `getBatchRange` implementation (untouched from your prototype) returns a range where the `stopTime` is set to the start time of the next batch. This means that there's a long gap after the last layout and passive effect for that batch.

For example, given this:

![image](https://user-images.githubusercontent.com/12784593/90654339-de2f1c80-e272-11ea-9a1e-b70b8c572a96.png)

For the hovered batch, `getBatchRange`'s `stopTime` will be where the black bar is, not the green bar. Is this correct?

There's a more extreme example at https://github.com/MLH-Fellowship/scheduling-profiler-prototype/issues/124. This screenshot was created by clicking the zoom to batch button on the hovered measures:

![image](https://user-images.githubusercontent.com/12784593/89553912-b9818080-d840-11ea-938f-9a184c71abe4.png)

bvaughn replied:
Yeah that's not surprising :smile: I'm just not sure I agree with it anymore. 😆 

> Could I clarify what you mean by delayed passive effects? Are these passive effects that are executed some time after the commit is complete?

Passive effects are run async, next tick/frame, in most cases.

If another render needs to start before passive effects have been flushed, we'll just-in-time flush them before starting on the render work. So I think what you're showing above is that happening.

Although it looks odd, the passive effects *are* part of the work though. And they could be *slow* if you're doing something expensive, so I think it's important for us to include them in the zoomed-in phase.

taneliang commented on packages/react-devtools-scheduling-profiler/src/view-base/geometry.js:
```diff
@@ -0,0 +1,129 @@
+/**
+ * Prerequisite: rect1 must intersect with rect2.
+ */
+export function rectIntersectionWithRect(rect1: Rect, rect2: Rect): Rect {
```
Makes sense! When I wrote this I didn't want to call it `intersectionOfRects` because I thought it'd be more confusing that `unionOfRects` accepts any number of rects whereas this only accepts 2. But I agree that `intersectionOfRects` will be easier to read, especially since it's often called close to `rectIntersectsRect`. I'll rename it

taneliang commented on packages/react-devtools-scheduling-profiler/src/view-base/layouter.js:
```diff
@@ -0,0 +1,269 @@
+) => Layout;
+
+function viewToLayoutInfo(view: View): LayoutInfo {
+  return {view, frame: view.frame};
```
> This conversion seems not very useful. Can we get rid of it?

I don't think we can 😆

Because the layouters are composable (via `createComposedLayout`), they need some way to pass the intermediate frames to each other. If we get rid of this `LayoutInfo` wrapper and store the intermediate frames in the views, we have these problems:

1. Invalidated views are redrawn the next time `CanvasPage`'s layout effects are flushed. Because the `View` class's `setFrame` method invalidates the view's contents (by calling `setsNeedsDisplay`) if the frame is changed, we only want to call `setFrame` with the final frame, so that the view will only be redrawn if the frame actually changed.

	Without the `LayoutInfo` wrapper, the intermediate frames will trigger a lot of unnecessary invalidation and thus many views will be unnecessarily redrawn on every frame.

1. View subclasses can override the `setFrame` method and do arbitrary things, including updating its state. For example, `HorizontalPanAndZoomView` and `VerticalScrollView` re-clamp their offsets and zoom levels when the frame changes, so that e.g. resizing the browser window don't cause black areas to appear. If this is done with intermediate frames, this clamping can cause bugs, e.g. if an intermediate frame had a much wider `width` than the final one, resizing the browser window may cause you to suddenly scroll left as the offset got clamped with that intermediate `width`.

This `LayoutInfo` wrapper was the best way I could think of to solve this. Is there a better way to solve this?

taneliang commented on packages/react-devtools-scheduling-profiler/src/context/Contexts.js:
```diff
@@ -0,0 +1,84 @@
+  onChange?: OnChangeFn,
+  pageX: number,
+  pageY: number,
+|}) {
```
Ah, nice catch! I'll do that.

Should I also make this change in `react-devtools-shared`'s implementation too? Also, would it be a good idea to merge the 2 context menu implementations after this PR? I think the only differences are that the scheduling profiler's implementation has an extra `onChange` callback (from your prototype) and slightly more comprehensive Flow types (which I added), as well as some CSS changes.

bvaughn replied:
Yeah if the same bug exists in `react-devtools-shared'` it would be nice to do a fix for that as well (maybe as a separate PR tho :grin:)

Let's not do any code-sharing yet.

bvaughn commented on packages/react-devtools-scheduling-profiler/src/CanvasPage.js:
```diff
@@ -269,22 +269,19 @@ function AutoSizedCanvas({data, height, width}: AutoSizedCanvasProps) {
+      x: interaction.payload.event.x,
+      y: interaction.payload.event.y,
+    });
+  });
```
We should still pass a deps array to `useCallback`. That array should just be empty.

Without the deps array, `useCallback` will just always pass the new function every time it's called.

bvaughn commented on packages/react-devtools-scheduling-profiler/src/content-views/UserTimingMarksView.js:
```diff
@@ -0,0 +1,230 @@
+  /**
+   * Draw a single `UserTimingMark` as a circle in the canvas.
+   */
+  _drawSingleMark(
+    context: CanvasRenderingContext2D,
+    rect: Rect,
+    mark: UserTimingMark,
+    baseY: number,
+    scaleFactor: number,
+    showHoverHighlight: boolean,
+  ) {
+    const {frame} = this;
+    const {timestamp} = mark;
+
+    const x = timestampToPosition(timestamp, scaleFactor, frame);
+    const radius = EVENT_DIAMETER / 2;
+    const markRect: Rect = {
+      origin: {
+        x: x - radius,
+        y: baseY,
+      },
+      size: {width: EVENT_DIAMETER, height: EVENT_DIAMETER},
+    };
+    if (!rectIntersectsRect(markRect, rect)) {
+      return; // Not in view
+    }
+
+    const fillStyle = showHoverHighlight
+      ? COLORS.USER_TIMING_HOVER
+      : COLORS.USER_TIMING;
+
+    if (fillStyle !== null) {
+      const y = markRect.origin.y + radius;
+
+      context.beginPath();
+      context.fillStyle = fillStyle;
+      context.arc(x, y, radius, 0, 2 * Math.PI);
+      context.fill();
+    }
+  }
```
Can we give these marks a different visual style, to set them apart from the e.g. state updates and suspends? Maybe something like:
```suggestion
  _drawSingleMark(
    context: CanvasRenderingContext2D,
    rect: Rect,
    mark: UserTimingMark,
    baseY: number,
    scaleFactor: number,
    showHoverHighlight: boolean,
  ) {
    const {frame} = this;
    const {timestamp} = mark;

    const x = timestampToPosition(timestamp, scaleFactor, frame);
    const size = USER_TIMING_MARK_SIZE;
    const halfSize = size / 2;

    const markRect: Rect = {
      origin: {
        x: x - halfSize,
        y: baseY,
      },
      size: {width: size, height: size},
    };
    if (!rectIntersectsRect(markRect, rect)) {
      return; // Not in view
    }

    const fillStyle = showHoverHighlight
      ? COLORS.USER_TIMING_HOVER
      : COLORS.USER_TIMING;

    if (fillStyle !== null) {
      const y = baseY + halfSize;

      context.beginPath();
      context.fillStyle = fillStyle;
      context.moveTo(x, y - halfSize);
      context.lineTo(x + halfSize, y);
      context.lineTo(x, y + halfSize);
      context.lineTo(x - halfSize, y);
      context.fill();
    }
  }
```

The above proposal also involves adding a new constant to `content-views/constants.js`:
```js
export const USER_TIMING_MARK_SIZE = 8;
```

bvaughn commented:
I think this is okay to merge as-is. I have some small suggestions (mentioned above) that could be made in a follow up PR, either by myself or you folks.

I want to land #19647 first though, on its own. Then we can land this one.

kartikcho commented on packages/react-devtools-scheduling-profiler/src/ImportPage.js:
```diff
@@ -0,0 +1,119 @@
+                  <a href="https://github.com/MLH-Fellowship/scheduling-profiler-prototype">
+                    <button className={style.ViewSourceButton}>
+                      <span>Source </span>
+                    </button>
```
@bvaughn the new import page that you saw in the meeting was a WIP [PR](https://github.com/MLH-Fellowship/scheduling-profiler-prototype/pull/135) that is on hold until this is merged. I'll be making a PR here soon after this is merged to get that in!

bvaughn replied:
Ah! I mixed PRs up then.

Sounds good. Thanks @kartikcho

bvaughn commented:
Looks like the `yarn.lock` file now conflicts.

bvaughn commented:
I don't have permissions to push to this repo, so I can't fix it. @kartikcho can you merge master and resolve the conflict?

taneliang commented:
> I think this is okay to merge as-is.

Exciting! 🥳

> I have some small suggestions (mentioned above) that could be made in a follow up PR, either by myself or you folks.

Thanks for the suggestions! I'll look at them in greater detail tomorrow as it's late here, but I really like the idea of having a `ViewManager` 😄

bvaughn commented:
Excellent work @taneliang and @kartikcho!

I'm going to merge this now and we/I can follow up on some of the remaining items afterward.

taneliang commented on packages/react-devtools-scheduling-profiler/src/view-base/ResizableSplitView.js:
```diff
@@ -0,0 +1,311 @@
+  }
+
+  _getResizeBar(): View {
+    return this.subviews[1];
```
Yep, I actually implemented something similar initially. Unfortunately, I didn't think it was very clean as it forced all view subclasses to correctly handle and maintain subview/superview relationships. I figured it'll be better to centralize all of that in the base `View` class.

Regarding this particular issue, maybe it'll be better if we stored a reference to the views in the subclass? It's definitely still not great though. Seems like there are some deep seated issues in this architecture 😞 

```js
export class ResizableSplitView extends View {
  // No casting or Array access necessary:
  _bottomSubview: View;
  _resizeBar: View;
  _topSubview: View;

  constructor(
    surface: Surface,
    frame: Rect,
    topSubview: View,
    bottomSubview: View
  ) {
    super(surface, frame, noopLayout);

    this._bottomSubview = bottomSubview;
    this._resizeBar = new ResizeBar(surface, frame);
    this._topSubview = topSubview;

    this.addSubview(this._topSubview);
    this.addSubview(this._resizeBar);
    this.addSubview(this._bottomSubview);

    // ...
  }
  // ...
}
```

taneliang commented on packages/react-devtools-scheduling-profiler/src/view-base/View.js:
```diff
@@ -0,0 +1,274 @@
+    this.subviews.forEach(subview =>
+      subview.handleInteractionAndPropagateToSubviews(interaction),
+    );
+  }
```
Hmm, maybe I'm not understanding your `ViewManager` design correctly. If the `ViewManager` class is an implementation detail of the `View` class, won't `View` still need to implement all its existing methods, including `handleInteractionAndPropagateToSubviews`?

What about this alternative design:

1. Turn `View` into an interface that only implements e.g. `draw`, `handleInteraction`, etc.
1. Define a new *public* `ViewManager` class that handles the invalidation and interaction propagation logic. 

`CanvasPage` will then instantiate `ViewManager`s and inject concrete `View` implementations into them.

This will mean that the manager will handle subViewManager/superViewManager relationships. Also, `View` implementations that instantiate their own subviews (e.g. `ResizableSplitView`, `FlamechartView`) will need to know how to instantiate `ViewManager`s, which doesn't feel right.
